File: JASSjr_index.java
/*
	JASSJR_INDEX.JAVA
	----------------
	Copyright (c) 2019 Andrew Trotman and Kat Lilly
	Minimalistic BM25 search engine.
*/
import java.util.HashMap;
import java.util.ArrayList;
import java.util.stream.Stream;
import java.nio.file.Files;
import java.io.IOException;
import java.nio.file.Paths;
import java.lang.Thread;
import java.io.FileOutputStream;
import java.io.DataOutputStream;


class JASSjr_index
{ 
    public class Posting 
    {
        public int d, tf;
	
	Posting(int d, int tf)
	{
	    this.d = d;
	    this.tf = tf;
	}
    }
    
    String buffer;
    int current;
    String next_token;
    HashMap<String, ArrayList<Posting>> vocab = new HashMap<String, ArrayList<Posting>>();
    ArrayList<String> doc_ids = new ArrayList<String>();
    ArrayList<Integer> length_vector = new ArrayList<Integer>();

    /* 
      TO_LITTLE()
      -----------
      Rearrange byte order so index matches that of CPP indexer
    */
    public int to_little(int value)
    {
	return ((value & 0xFF) << 24) | (((value >>> 8) & 0xFF) << 16) | (((value >>> 16) & 0xFF) << 8) | (((value >>> 14) & 0xFF) << 0);
    }
    
    /*
      LEX_GET_NEXT()
      --------------
      One-character lookahead lexical analyser
    */
    public String lex_get_next()
    {
	/*
	  Skip over whitespace and punctuation (but not XML tags)
	*/	
	while (current < buffer.length() && !Character.isLetterOrDigit(buffer.charAt(current)) && buffer.charAt(current) != '<')
	    current++;

	/*
	  A token is either an XML tag '<'..'>' or a sequence of alpha-numerics.
	*/
	int start = current;
	if (current >= buffer.length())
	    return null;     // must be at end of line
	else if (Character.isLetterOrDigit(buffer.charAt(current)))
	    while (current < buffer.length() && (Character.isLetterOrDigit(buffer.charAt(current)) || buffer.charAt(current) == '-'))				// TREC <DOCNO> primary keys have a hyphen in them
		current++;
	else if (buffer.charAt(current) == '<')
	{
	    current++;
	    while (current < buffer.length() && buffer.charAt(current - 1) != '>')
		current++;
	}
	/*
	  Copy and return the token
	*/		
	return buffer.substring(start, current);
    }
    
    /*
      LEX_GET_FIRST()
      ---------------
      Start the lexical analysis process
    */
    public String lex_get_first(String with)
    {
	buffer = with;
	current = 0;

	return lex_get_next();
    }

    /*
	GO()
	----
	Simple indexer for TREC WSJ collection
    */
    public void go(String args[]) 
    {
	int docid = -1;
	int document_length = 0;

	/*
	  Make sure we have one paramter, the filename
	*/
	if (args.length != 1)
	{
	    System.out.println("Usage: java " + Thread.currentThread().getStackTrace()[1].getClassName() + " <infile.xml>");
	    System.exit(0);
	}

	try (Stream<String> stream = Files.lines(Paths.get(args[0])))
	{
	    for (String line : (Iterable<String>) stream::iterator)
	    {
		String token;
		Boolean push_next = false;
		for (token = lex_get_first(line); token != null; token = lex_get_next())
		{
		    if (token.equals("<DOC>"))
		    {
			/*
			  Save the previous document length
			*/
			if (docid != -1)
			    length_vector.add(document_length);
			
			/*
			  Move on to the next document
			*/
			docid++;
			document_length = 0;
			
			if ((docid % 1000) == 0)
			    System.out.println(docid + " documents indexed");
		    }

		    /*
		      if the last token we saw was a <DOCID> then the next token is the primary key
		    */
		    if (push_next)
		    {
			doc_ids.add(token);
			push_next = false;
		    }
		    
		    if (token.equals("<DOCNO>"))
			push_next = true;
		    
		    /*
		      Don't index XML tags
		    */
		    if (token.charAt(0) == '<')
			continue;

		    /*
		      lower case the string
		    */
		    token.toLowerCase();

		    /*
		      truncate any long tokens at 255 charactes (so that the length can be stored first and in a single byte)
		    */
		    if (token.length() > 0xFF)
			token = token.substring(0, 0xFF);

		    /*
		      add the posting to the in-memory index
		    */
		    ArrayList<Posting> list = vocab.get(token);
		    if (list == null)
		    {
			ArrayList<Posting> new_list = new ArrayList<Posting>();
			new_list.add(new Posting(docid, 1));
			vocab.put(token, new_list);     // if the term isn't in the vocab yet 
		    }
		    else
		    {
			if (list.get(list.size() - 1).d != docid)
			    list.add(new Posting(docid, 1));							// if the docno for this occurence hasn't changed the increase tf
			else
			    list.get(list.size() - 1).tf++;												// else create a new <d,tf> pair.
		    }
		    
		    /*
		      Compute the document length
		    */
		    document_length++;
		}
	    }
	}
	catch (IOException e)
	{
	    e.printStackTrace();
	}
	
	/*
	  tell the user we've got to the end of parsing
	*/
	System.out.println("Indexed " + (docid + 1) + " documents. Serialising...");
	
	/*
	  Save the final document length
	*/
	length_vector.add(document_length);

	try
	{
	    /*
	      store the primary keys
	    */
	    DataOutputStream docids_out = new DataOutputStream(new FileOutputStream("docids.bin"));
	    for (int i= 0; i < doc_ids.size(); i++)
	    {
		docids_out.write(doc_ids.get(i).getBytes(), 0, (byte) doc_ids.get(i).length());
		docids_out.write('\n');
	    }
	    
	    /*
	      serialise the in-memory index to disk
	    */    
	    FileOutputStream postings_file = new FileOutputStream("postings.bin");
	    DataOutputStream postings_out = new DataOutputStream(postings_file);
	    DataOutputStream vocab_out = new DataOutputStream(new FileOutputStream("vocab.bin"));
	    
	    for (HashMap.Entry<String, ArrayList<Posting>> entry : vocab.entrySet())
	    {
		/*
		  write the postings list to one file
		*/
		int where = (int) postings_file.getChannel().position();
		    
		for (Posting pair : entry.getValue())
		{
		    postings_out.writeInt(to_little(pair.d));
		    postings_out.writeInt(to_little(pair.tf));
		}
		    
		/*
		  write the vocabulary to a second file (one byte length, string, '\0', 4 byte where, 4 byte size)
		*/
		vocab_out.write((byte) entry.getKey().length());
		vocab_out.write(entry.getKey().getBytes(), 0, (byte) entry.getKey().length());
		vocab_out.write('\0');
		vocab_out.writeInt(to_little(where));
		vocab_out.writeInt(to_little((int) postings_file.getChannel().position() - where));
	    }

	    /*
	      store the document lengths
	    */
	    DataOutputStream doclengths_out = new DataOutputStream(new FileOutputStream("lengths.bin"));
	    for (int i= 0; i < length_vector.size(); i++)
		doclengths_out.writeInt(to_little(length_vector.get(i)));
		   
	    /*
	      clean up
	    */
	    docids_out.close();
	    postings_out.close();
	    vocab_out.close();
	    doclengths_out.close();
	}
	catch (Exception e)
	{
	    System.out.println(e);
	    e.printStackTrace();
	}
    }
		  
    public static void main(String args[]) 
    {
	JASSjr_index indexer = new JASSjr_index();
	indexer.go(args);
    } 
} 

File: JASSjr_index.zig
// Copyright (c) 2024, 2026 Vaughan Kitchen
// Minimalistic BM25 search engine.

const std = @import("std");
const native_endian = @import("builtin").target.cpu.arch.endian();

// Pair definition for a value in a postings list
const Posting = struct { i32, i32 };

// One-character lookahead lexical analyser
const Lexer = struct {
    buffer: []u8,
    current: usize,

    fn init(buffer: []u8) Lexer {
        return Lexer{ .buffer = buffer, .current = 0 };
    }

    // Conform to Zig naming conventions for iterators
    fn next(self: *Lexer) ?[]u8 {
        // Skip over whitespace and punctuation (but not XML tags)
        while (self.current < self.buffer.len and !std.ascii.isAlphanumeric(self.buffer[self.current]) and self.buffer[self.current] != '<')
            self.current += 1;

        // A token is either an XML tag '<'..'>' or a sequence of alpha-numerics.
        const start = self.current;
        if (self.current >= self.buffer.len) {
            return null; // must be at end of line
        } else if (std.ascii.isAlphanumeric(self.buffer[self.current])) {
            // TREC <DOCNO> primary keys have a hyphen in them
            while (self.current < self.buffer.len and (std.ascii.isAlphanumeric(self.buffer[self.current]) or self.buffer[self.current] == '-'))
                self.current += 1;
        } else if (self.buffer[self.current] == '<') {
            self.current += 1;
            while (self.current < self.buffer.len and self.buffer[self.current - 1] != '>')
                self.current += 1;
        }

        // Return the token as a slice. This won't live past the underlying buffer
        return self.buffer[start..self.current];
    }
};

// Simple indexer for TREC WSJ collection
pub fn main(init: std.process.Init) !void {
    var arena = init.arena.allocator();
    var stdout_buffer: [1024]u8 = undefined;
    var stdout_writer = std.Io.File.stdout().writer(init.io, &stdout_buffer);
    const stdout = &stdout_writer.interface;

    const argv = try init.minimal.args.toSlice(arena);

    // Make sure we have one parameter, the filename
    if (argv.len != 2) {
        try stdout.print("Usage: {s} <infile.xml>\n", .{argv[0]});
        try stdout_writer.flush();
        std.process.exit(0);
    }

    var vocab = std.StringHashMap(std.ArrayList(Posting)).init(arena);
    var doc_ids: std.ArrayList([]u8) = .empty;
    var doc_lengths: std.ArrayList(i32) = .empty;

    var doc_id: i32 = -1;
    var document_length: i32 = 0;

    var fh_buffer: [2048]u8 = undefined;
    var fh = try std.Io.Dir.cwd().openFile(init.io, argv[1], .{});
    var stream = fh.reader(init.io, &fh_buffer);

    var push_next = false;
    while (try stream.interface.takeDelimiter('\n')) |line| {
        var lex = Lexer.init(line);
        while (lex.next()) |token| {
            if (std.mem.eql(u8, token, "<DOC>")) {
                // Save the previous document length
                if (doc_id != -1)
                    try doc_lengths.append(arena, document_length);
                // Move on to the next document
                doc_id += 1;
                document_length = 0;
                if (@rem(doc_id, 1000) == 0) {
                    try stdout.print("{d} documents indexed\n", .{doc_id});
                    try stdout_writer.flush();
                }
            }
            // If the last token we saw was a <DOCNO> then the next token is the primary key
            if (push_next) {
                const primary_key = try arena.dupe(u8, token);
                try doc_ids.append(arena, primary_key);
                push_next = false;
            }
            if (std.mem.eql(u8, token, "<DOCNO>")) {
                push_next = true;
            }
            // Don't index XML tags
            if (token[0] == '<')
                continue;

            // Lower case the string
            _ = std.ascii.lowerString(token, token);

            // Truncate any long tokens at 255 characters (so that the length can be stored first and in a single byte)
            const token2 = if (token.len < 255) token else token[0..255];

            // Add the posting to the in-memory index
            const gop = try vocab.getOrPut(token2);
            if (!gop.found_existing) {
                // If the term isn't in the vocab yet
                const term = try arena.dupe(u8, token2);
                gop.key_ptr.* = term;
                gop.value_ptr.* = .empty;
                try gop.value_ptr.append(arena, .{ doc_id, 1 });
            } else {
                if (gop.value_ptr.getLast()[0] != doc_id) {
                    // If the docno for this occurence has changed then create a new <d,tf> pair
                    try gop.value_ptr.append(arena, .{ doc_id, 1 });
                } else {
                    // Else increase the tf
                    gop.value_ptr.items[gop.value_ptr.items.len - 1][1] += 1;
                }
            }

            // Compute the document length
            document_length += 1;
        }
    }

    // Save the final document length
    try doc_lengths.append(arena, document_length);

    // Tell the user we've got to the end of parsing
    try stdout.print("Indexed {d} documents. Serialising...\n", .{doc_id + 1});
    try stdout_writer.flush();

    // Store the primary keys
    var docids_buffer: [1024]u8 = undefined;
    const docids_fh = try std.Io.Dir.cwd().createFile(init.io, "docids.bin", .{});
    var docids_stream = docids_fh.writer(init.io, &docids_buffer);

    for (doc_ids.items) |primary_key| {
        try docids_stream.interface.writeAll(primary_key);
        try docids_stream.interface.writeByte('\n');
    }

    // Serialise the in-memory index to disk
    var postings_buffer: [1024]u8 = undefined;
    const postings_fh = try std.Io.Dir.cwd().createFile(init.io, "postings.bin", .{});
    var postings_stream = postings_fh.writer(init.io, &postings_buffer);

    var vocab_buffer: [1024]u8 = undefined;
    const vocab_fh = try std.Io.Dir.cwd().createFile(init.io, "vocab.bin", .{});
    var vocab_stream = vocab_fh.writer(init.io, &vocab_buffer);

    var where: usize = 0;
    var it = vocab.iterator();
    while (it.next()) |kv| {
        // Write the postings list to one file
        try postings_stream.interface.writeAll(std.mem.sliceAsBytes(kv.value_ptr.items));

        // Write the vocabulary to a second file (one byte length, string, '\0', 4 byte where, 4 byte size)
        try vocab_stream.interface.writeByte(@truncate(kv.key_ptr.len));
        try vocab_stream.interface.writeAll(kv.key_ptr.*);
        try vocab_stream.interface.writeByte(0);
        try vocab_stream.interface.writeInt(u32, @truncate(where), native_endian);
        try vocab_stream.interface.writeInt(u32, @truncate(kv.value_ptr.items.len * 8), native_endian);

        where += kv.value_ptr.items.len * 8;
    }

    // Store the document lengths
    var lengths_buffer: [1024]u8 = undefined;
    const lengths_fh = try std.Io.Dir.cwd().createFile(init.io, "lengths.bin", .{});
    var lengths_stream = lengths_fh.writer(init.io, &lengths_buffer);
    try lengths_stream.interface.writeAll(std.mem.sliceAsBytes(doc_lengths.items));

    // Cleanup
    try lengths_stream.flush();
    lengths_fh.close(init.io);

    try vocab_stream.flush();
    vocab_fh.close(init.io);

    try postings_stream.flush();
    postings_fh.close(init.io);

    try docids_stream.flush();
    docids_fh.close(init.io);
}
